src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -21,180 +21,168 @@ internal sealed class TaskCenterSolutionAnalysisProgressReporter
-                _resettableDelay.Task.SafeContinueWith(_ =>
+                // Kick off task to update the UI after a delay to pick up any new events.
+                _intervalTask = Task.Delay(s_minimumInterval).ContinueWith(_ =>
                 {
-                    _resettableDelay = null;
-                    ProgressUpdated();
+                    ReportProgress();
                 }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
             }
         }
 
-        private void StartedOrStopped(bool started)
+        private void ReportProgress()
         {
-            if (started)
+            ProgressData data;
+            lock (_lock)
             {
-                ResetProgressStatus();
+                data = _lastProgressData;
+                _intervalTask = null;
 
-                // if there is any pending one. make sure it is finished.
-                _currentTask?.TrySetResult(default);
-
-                var taskHandler = _taskCenterService.PreRegister(_options, data: default);
-
-                _currentTask = new TaskCompletionSource<VoidResult>();
-                taskHandler.RegisterTask(_currentTask.Task);
+                _updateUITask = _updateUITask.ContinueWith(_ => UpdateUI(data),
+                    CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+            }
+        }
 
-                // report initial progress
-                ChangeProgress(taskHandler, message: null);
+        private void UpdateUI(ProgressData progressData)
+        {
+            if (progressData.Status == ProgressStatus.Stopped)
+            {
+                StopTaskCenter();
+                return;
+            }
 
-                // set handler
-                _taskHandler = taskHandler;
+            // Update the pending item count if the progress data specifies a value.
+            if (progressData.PendingItemCount.HasValue)
+            {
+                _lastPendingItemCount = progressData.PendingItemCount.Value;
             }
-            else
+
+            // Start the task center task if not already running.
+            if (_taskHandler == null)
             {
-                // clear progress message
-                ChangeProgress(_taskHandler, message: null);
+                StartTaskCenter();
+            }
 
-                // stop progress
-                _currentTask?.TrySetResult(default);
-                _currentTask = null;
+            var statusMessage = progressData.Status == ProgressStatus.Paused
+                ? ServicesVSResources.Paused_0_tasks_in_queue
+                : ServicesVSResources.Evaluating_0_tasks_in_queue;
+            ChangeProgress(string.Format(statusMessage, _lastPendingItemCount));
+        }
 
-                _taskHandler = null;
+        private void StartTaskCenter()
```

Comment:
up to you, but you could conisider inlining this.

Reply:
I kinda prefer it outside with separate start stop methods. seems to make it easier to read for me so I'll leave it as is.

Reply:
👍